resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -1277,10 +1277,14 @@ private[spark] class Client(
     } else {
       val YarnAppReport(appState, finalState, diags) = monitorApplication(appId)
       if (appState == YarnApplicationState.FAILED || finalState == FinalApplicationStatus.FAILED) {
+        var amContainerSuccess = false
         diags.foreach { err =>
+          amContainerSuccess = err.contains("AM Container") && err.contains("exitCode: 0")
           logError(s"Application diagnostics message: $err")
         }
-        throw new SparkException(s"Application $appId finished with failed status")
+        if (!amContainerSuccess) {
+          throw new SparkException(s"Application $appId finished with failed status")
+        }
```

Comment:
Instead of parsing this from diagnostics, we can use AM's getContainerExitStatus and check for that being 0 ?

Reply:
> getDriverLogsLink

Nice suggestion! updated

Reply:
> > getDriverLogsLink
> 
> Nice suggestion! updated

Tested in our prod,  throw such exception
```
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.yarn.exceptions.ContainerNotFoundException): Container with id 'container_e240_1632999510150_3975819_02_000001' doesn't exist in RM.
	at org.apache.hadoop.yarn.server.resourcemanager.ClientRMService.getContainerReport(ClientRMService.java:532)
	at org.apache.hadoop.yarn.api.impl.pb.service.ApplicationClientProtocolPBServiceImpl.getContainerReport(ApplicationClientProtocolPBServiceImpl.java:466)
	at org.apache.hadoop.yarn.proto.ApplicationClientProtocol$ApplicationClientProtocolService$2.callBlockingMethod(ApplicationClientProtocol.java:645)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:622)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:590)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:574)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1129)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1056)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:984)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1972)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3039)
```

Reply:
> Instead of parsing this from diagnostics, we can use AM's getContainerExitStatus and check for that being 0 ? Take a look at `getDriverLogsLink` on how to get to AM container

Tested in our prod and check yarn's code

![image](https://user-images.githubusercontent.com/46485123/140268608-c93b4f90-0d99-43be-97c0-d66ed4f49169.png)

RM will directly delete application attempts when done application.
And throw exception when  use current code
```
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.yarn.exceptions.ContainerNotFoundException): Container with id 'container_e240_1632999510150_3975819_02_000001' doesn't exist in RM.
	at org.apache.hadoop.yarn.server.resourcemanager.ClientRMService.getContainerReport(ClientRMService.java:532)
	at org.apache.hadoop.yarn.api.impl.pb.service.ApplicationClientProtocolPBServiceImpl.getContainerReport(ApplicationClientProtocolPBServiceImpl.java:466)
	at org.apache.hadoop.yarn.proto.ApplicationClientProtocol$ApplicationClientProtocolService$2.callBlockingMethod(ApplicationClientProtocol.java:645)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:622)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:590)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:574)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1129)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1056)
```

May need to extract from diagnostics...

Reply:
Thanks for the details @AngersZhuuuu, this is unfortunate.
I wanted to avoid parsing from diagnostic, given it is potentially brittle - without timeline service, I am not sure how to extract this without parsing the string.
And thoughts @tgravescs ?

Reply:
yeah unfortunately that final application status is what YARN really uses and advertises and it really wants you to unregister.

How often does this really happen?  it seems like if you are timing out talking to RM very often you have other problems on your cluster.  Or does this happen on rolling upgrade or something?

Note, If we were to do this, then the YARN final status also wouldn't match because doesn't that show up as failed when it can't unregister?  that could confuse the user.

Reply:
> Thanks for the details @AngersZhuuuu, this is unfortunate. I wanted to avoid parsing from diagnostic, given it is potentially brittle - without timeline service, I am not sure how to extract this without parsing the string. Any thoughts @tgravescs ?

Yea, we can got this from timeline service, but it's also strange..

Reply:
> yeah unfortunately that final application status is what YARN really uses and advertises and it really wants you to unregister.
>
> How often does this really happen? it seems like if you are timing out talking to RM very often you have other problems on your cluster. Or does this happen on rolling upgrade or something?

This situation does not happen often, but it will happen from time to time, causing many users to complain. Our cluster is relatively large, and the performance pressure of yarn is very high, so sometimes, some requests will respond very slowly. Then caused such issue.


> Note, If we were to do this, then the YARN final status also wouldn't match because doesn't that show up as failed when it can't unregister? that could confuse the user.

People always care application's running status more than the yarn's info. If job success, they won't check applications on yarn ui. For yarn's admin,  The information for this job is still the same as before


Also we can add a legacy configuration for this

Reply:
Users complain because miss SLA or something?  Retry should be an acceptable thing sometimes when failures happen.  But I get it that its annoying.

What is the timeout like in this case?  Is it the YARN call timed out but would have answered within a few 
seconds or is it YARN isn't going to respond for 10's of minutes or hours?   

The other thing we could possibly do is look at adding optional retry logic in Spark on top of YARN client retry that is configurable where we say we really want to unregister, so wait and retry in addition to the built in Hadoop retry logic.  That would delay the client knowing its done but would NOT retry and would keep YARN state consistent.  Obviously would need others input on this as well.   This is one reason why I'm asking about when this happens and exact circumstances so that we could possibly come up with alternate solutions.

Like @mridulm  I'm not a fan of parsing the diagnostics.

you can also increase your rpc timeout for talking to the resource manager. 

the other way is to handle this outside of Spark.. application has retry logic instead of spark where it can check to see if for instance it wrote the expected output and if its there don't retry, otherwise retry.

Reply:
> Users complain because miss SLA or something? Retry should be an acceptable thing sometimes when failures happen. But I get it that its annoying.

Yea, really annoying and miss SLA


> What is the timeout like in this case? Is it the YARN call timed out but would have answered within a few seconds or is it YARN isn't going to respond for 10's of minutes or hours?

There is a shutdown hook timeout of 30s.

> The other thing we could possibly do is look at adding optional retry logic in Spark on top of YARN client retry that is configurable where we say we really want to unregister, so wait and retry in addition to the built in Hadoop retry logic. That would delay the client knowing its done but would NOT retry and would keep YARN state consistent. Obviously would need others input on this as well. This is one reason why I'm asking about when this happens and exact circumstances so that we could possibly come up with alternate solutions.

Still need retry. The best way should be handling such case in yarn side, add an optional logic that when AM container exitcode 0, return SUCCESS final status.

Reply:
> There is a shutdown hook timeout of 30s.

This doesn't answer my question.  How long does it take for YARN to actually respond in the cases this is slow?

The way YARN API is supposed to work, the application master is meant to unregister, it is not meant to go off of exit codes, if you can't unregister, YARN doesn't really know what happened.  Yes perhaps YARN should expose container exit status better, feel free to go propose a better API in YARN that we can use.

honestly this is kind of a YARN cluster setup issue in my opinion, if your ResourceManager is not responding for long periods of times frequently, you need to fix that.  It would be great if applications could work around that but I don't want to put something into Spark that is brittle and could cause other issues.

Reply:
> > There is a shutdown hook timeout of 30s.
> 
> This doesn't answer my question. How long does it take for YARN to actually respond in the cases this is slow?
> 
> The way YARN API is supposed to work, the application master is meant to unregister, it is not meant to go off of exit codes, if you can't unregister, YARN doesn't really know what happened. Yes perhaps YARN should expose container exit status better, feel free to go propose a better API in YARN that we can use.
> 
> honestly this is kind of a YARN cluster setup issue in my opinion, if your ResourceManager is not responding for long periods of times frequently, you need to fix that. It would be great if applications could work around that but I don't want to put something into Spark that is brittle and could cause other issues.

In the previous yarn performance degradation, this situation occurred. That time, you can check the RPC response time by opening the debug log, but this time it was an occasional situation, so it was impossible to test. Previously, there was a case where yarn accepted the unregister request for more than 30 seconds and did not return.

Reply:
so I guess I'm kind of wondering if we should do make any change here.
This doesn't fail the application, it may be unwanted behavior in your case that it retries sometimes but if its doing it to often it seems like a cluster issue to me.  Hadoop client has built in retries and timeouts that are supposed to be configured to account for some of these, so if you are going outside of those that should be a out of the norm occurrence.  The Hadoop philosophy IMHO is to retry on failures so if this occasionally happens shouldn't be a big deal.  If it is, it's up to the application to handle. The Hadoop API for properly checking application status is what we are using and this trying to infer the status which could be brittle and just adds maintenance for Spark if we add yet another config.   

Note Hadoop also has a shutdown hook configuration which I think is the 30 seconds you are referring to here: `hadoop.service.shutdown.timeout`.   You could increase that and increase the Hadoop retries/timeouts.   

It would be nice to get more feedback on if this is a problem for other users or if devs have opinions.